[aotd] Fuse tangents subclasses runtime traversals

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #136478
* #139095
* __->__ #139068


Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

IvanKobzarev committed:
[aotd] Fuse tangents subclasses runtime traversals

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

bdhirsh commented:
I see some test failures?

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev commented:
> I see some test failures?

Yes, checking.

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -2463,6 +2466,7 @@ def fn(x: torch.Tensor, x1: torch.Tensor) -> torch.Tensor:
         # Not checking equality of ref and x as Exception is expected
 
     # Partially addresses https://github.com/pytorch/pytorch/issues/106457
+    @skipIfTorchDynamo()
```
it sounds like prior to this PR, this test would work properly under dynamo, but now it does not. Why?

bdhirsh replied:
Hmm if there answer is because dynamo blows up when trying to run directly on the new custom schema objects that we branch on at runtime, then I agree a skip here seems fine (it is unnecessary to get dynamo working on that). But I'd like a comment next to this @skip explaining exactly what we are not supporting in dynamo

IvanKobzarev replied:
There is an error on symbolic shapes guard verbose printing, that appeared after tangents processing change:

https://gist.github.com/IvanKobzarev/339f6b0b1465de56731cb6d6d14f2a9f

bdhirsh commented on test/functorch/test_aotdispatch.py:
```diff
@@ -6184,6 +6195,75 @@ def fn(x):
+    @unittest.skipIf(
+        not torch.distributed.is_available(), "test requires torch distributed"
+    )
+    @skipIfTorchDynamo()
```
nit: `test_dtensor_compile.py` is probably a better fit for this test:

(1) it's testing `AsyncCollectiveTensor`, which is more of a distributed concept

(2) then we won't need to worry about the `skipIfTorchDynamo` logic, since the tests in that file won't involve dynamo running on the AOTAutograd code.

IvanKobzarev replied:
Agree, moved to test_dtensor_compile

bdhirsh commented on torch/_functorch/_aot_autograd/subclass_utils.py:
```diff
@@ -70,33 +109,23 @@ def create_subclass_metadata(a: Any, start_idx: int, count_symints: bool):
     *,
     count_symints: bool = True,
-) -> List[Union[int, SubclassCreationMeta]]:
+    with_memory_format: bool = False,
```
Can you help me understand why we want to sometimes **not** include memory_format  when creating subclass meta? If there is a good reason for doing it sometimes and not others, a comment explaining exactly when it is / is not necessary would be nice.

IvanKobzarev replied:
My main logic was to not add overhead on deducing memory_format.
This  could also be especially painful if to call it during tracing on FakeTensors with symbolic shapes - memory format checks in my experience  give hairy symbolic shapes checks on strides (divisibility, equal to 1, reminder equals 0 etc.).

We use create_subclass_meta for inputs, outputs (in collect_metadata_analysis). I have not seen any usage of memory_format for inputs?

If we need memory format for inputs and outputs too - we can make it non-optional.

bdhirsh replied:
oh that's fair - we don't need the memory format info for inputs. Can you just mention that in a comment?

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1822,98 +1828,52 @@ def backward(ctx, *flat_args):
-                        (
-                            AOTDispatchAutograd.coerce_runtime_tangent(
+                    flat_processed_tangents = list(
+                        itertools.chain.from_iterable(
```
have you had a chance to benchmark if the runtime overhead here nets out to being faster/slower than the original code? (I'd imagine that merging the looping over tangents into a single loop would be faster, although I'm also not sure how `itertools.chain.from_iterable` is).

IvanKobzarev replied:
I measured itertools.chain.from_iterable vs sequential list.extend(), itertools.chain.from_iterable was insignificantly faster ( < 1%).

IvanKobzarev replied:
Using updated version of profiling PR https://github.com/pytorch/pytorch/pull/136478

Got that processing runtime tangents for recursive TwoTensor did not change (the diff in measurement std)

average before: 76610ns
average after: 76800ns

This of course depends how expensive is tensor_flatten call for SubClass, for TwoTensor it is cheap :)

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1441,55 +1442,83 @@ class AutogradLazyBackwardCompileInfo:
+    def process_runtime_tangent(x, meta: Union[PlainTensorMeta, SubclassCreationMeta]):
         if not isinstance(x, torch.Tensor):
-            return x
+            return x, [x]
```
I'm still trying to understand what the purpose of the second return argument of this function is. What do we need it for? (it looks like it's dropped in the outer-most call to `process_runtime_tangents`)

IvanKobzarev replied:
Current logic on tangents is:
```
tangents = all_args[TB, TE]
traverse_tangents_tree_to_check_type(tangents)
all_args = [traverse_subclass_tangents_coerce_metadata(all_args[i]) where i in [TB, TE]]
all_args = [traverse_subclass_tangents_coerce_memory_format(all_args[i]) where i in [TB, TE]]]
all_args = traverse_subclass_unwrap(all_args)
```
We are fusing all traversals that check/update in `process_runtime_tangents`,
and also we fuse `traverse_subclass_unwrap` into `process_runtime_tangents` doing flatenning at the same time of checks/updates. The second argument returns updated flattened leaves for each tangent.

As a result we come to the logic with only one subclasses tree traversal on runtime tangents, using second element in tuple as a result of unwrap.
```
processed_tangents = process_runtime_tangents(all_args[TB, TE])
processed_tangents_leaves = list(itertools.chain_from_iterable(pt[1]) for pt in processed_tangents)
all_args = traverse_subclass_unwrap(all_args[:TB]) + processed_tangents_leaves +  traverse_subclass_unwrap(all_args[TE+1:]) 
```

bdhirsh reviewed (commented):
looks mostly good - left a few comments

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



[ghstack-poisoned]

bdhirsh commented on torch/_functorch/_aot_autograd/runtime_wrappers.py:
```diff
@@ -1441,55 +1442,83 @@ class AutogradLazyBackwardCompileInfo:
 
-        return x
+        if is_traceable_wrapper_subclass(x):
+            runtime_meta = x.__tensor_flatten__()[1]
```
nit: I see we're calling `__tensor_flatten__()` twice, to get the metadata here and the inner keys later. If you think we can easily get away with a single call that seems better, but if not that's ok

IvanKobzarev replied:
Thanks. Yes, originally I thought that we should call __torch_flatten__ one more time after potential coercion (e.g. subclass type change) - I will make a check if x is unchanged - then we do not need extra tensor_flatten. But if coercion happened - than to call tensor_flatten.

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[aotd] Fuse tangents subclasses runtime traversals"



Reason:
Currently we have multiple traversals for tangents in runtime:
 - To check that types and structure are identical to what we guessed during tracing time
 - Coerce metadata
 - Coerce memory_format
 - Unwrap_tensor_subclass
All of them are traversing tangents via __tensor_flatten__ calls the tree of Subclasses.

Change:
To do everything in one traversal at runtime (including flattening)

Implementation details:

Add memory_format information inside SubclassCreationMeta, for PlainTensors keep not only (int) of unwrapped_index, but memory_format too.  

Preparing memory_format is optional (controlled by with_memory_format=True).

2. Removing unused subclass_utils.create_metadata_for_subclass which does not have any usages inside torch and would require update of the logic.



cc H-Huang awgu kwen2501 wanchaol fegin fduwjj wz337 wconstab d4l3k c-p-i-o

[ghstack-poisoned]

IvanKobzarev commented:
@pytorchbot merge
